server/chat-commands/avatars.tsx

```diff
@@ -0,0 +1,869 @@
+		`/avatars [username] - Shows custom avatars available to a user.`,
+		`!avatars - Show everyone that information. Requires: + % @ # &`,
+	],
+
+	addavatar() {
+		this.sendReply("Is this a personal avatar or a group avatar?");
+		return this.parse(`/help addavatar`);
+	},
+	addavatarhelp: [
+		`/defaultavatar [username], [avatar] - Gives a user a default (personal) avatar.`,
+		`/allowavatar [username], [avatar] - Gives a user an allowed (group) avatar.`,
+		`/removeavatar [username], [avatar] - Removes access to an avatar from a user.`,
+		`/removeavatar [username] - Removes access to all custom avatars from a user.`,
+		AVATAR_FORMATS_MESSAGE,
+	],
+
+	personalavatar: 'defaultavatar',
+	async defaultavatar(target, room, user) {
+		this.checkCan('bypassall');
+		if (!target) return this.parse(`/help defaultavatar`);
+		const [inputUsername, inputAvatar] = this.splitOne(target);
+		if (!Users.isUsername(inputUsername)) {
+			throw new Chat.ErrorMessage(`"${inputUsername}" is not a valid username.`);
+		}
+		const userid = toID(inputUsername);
+		const avatar = await Avatars.validate(inputAvatar, {rejectOfficial: true});
+
+		if (!Avatars.addPersonal(userid, avatar)) {
+			throw new Chat.ErrorMessage(`User "${inputUsername}" can already use avatar "${avatar}".`);
+		}
+		this.globalModlog('PAVATAR', userid, avatar);
+		this.sendReplyBox(<div>
+			{Avatars.img(avatar)}<br />
+			Added to <username class="username">{inputUsername}</username>
+		</div>);
+	},
+	defaultavatarhelp: 'addavatarhelp',
+
+	allowedavatar: 'allowavatar',
+	groupavatar: 'allowavatar',
+	async allowavatar(target, room, user) {
+		this.checkCan('bypassall');
+		if (!target) return this.parse(`/help defaultavatar`);
+		const [inputUsername, inputAvatar] = this.splitOne(target);
+		if (!Users.isUsername(inputUsername)) {
+			throw new Chat.ErrorMessage(`"${inputUsername}" is not a valid username.`);
+		}
+		const userid = toID(inputUsername);
+		const avatar = await Avatars.validate(inputAvatar, {rejectOfficial: true});
+
+		if (!Avatars.addAllowed(userid, avatar)) {
+			throw new Chat.ErrorMessage(`User "${inputUsername}" can already use avatar "${avatar}".`);
+		}
+		this.globalModlog('GAVATAR', userid, avatar);
+		this.sendReplyBox(<div>
+			{Avatars.img(avatar)}<br />
+			Added to <username class="username">{inputUsername}</username>
+		</div>);
+	},
+	allowavatarhelp: 'addavatarhelp',
+
+	denyavatar: 'removeavatar',
+	disallowavatar: 'removeavatar',
+	removeavatar(target, room, user) {
+		this.checkCan('bypassall');
+		if (!target) return this.parse(`/help defaultavatar`);
+		const [inputUsername, inputAvatar] = this.splitOne(target);
+		if (!Users.isUsername(inputUsername)) {
+			throw new Chat.ErrorMessage(`"${inputUsername}" is not a valid username.`);
+		}
+		const userid = toID(inputUsername);
+		const avatar = Avatars.convert(inputAvatar);
+
+		const allowed = customAvatars[userid]?.allowed.filter(Boolean);
+		if (!allowed) {
```

Comment:
```suggestion
		if (!allowed.length) {
```

Reply:
`allowed.filter(Boolean)` can never be empty.

Reply:
Hm, plausibly I could update the type to reflect this...

Reply:
Nah, that makes initialization hard. But the point is that if `allowed` were ever empty, `customAvatars[userid]` would just be deleted.

Reply:
But if it can never be empty, why do you check this? 
`!allowed` will never be true since [] is truthy.

Reply:
Oops. Misread.